lib/core/request.c

```diff
@@ -735,3 +736,22 @@ void h2o_resp_add_date_header(h2o_req_t *req)
     h2o_timestamp_t ts = h2o_get_timestamp(req->conn->ctx, &req->pool);
     h2o_add_header(&req->pool, &req->res.headers, H2O_TOKEN_DATE, NULL, ts.str->rfc1123, strlen(ts.str->rfc1123));
 }
+
+void h2o_send_early_hints(h2o_req_t *req)
+{
+    /* 103 must be sent before h2o_start_response is called*/
+    assert(req->_generator == NULL);
+    assert(req->_ostr_top->next == NULL && req->_ostr_top->send_early_hints != NULL);
```

Comment:
I think we should allow `req->_ostr_top->send_early_hints` being NULL for performance reasons, since for the most of the users H2O would not be emitting early hints for HTTP/1.

We can do `if (req->_ostr_top->send_early_hints == NULL) return;` here, and avoid the cost of running the header filters.

Reply:
so what do you think about how and where should we set `req->_ostr_top->send_early_hints`?

Reply:
You can continue setting up the `send_early_hints` callback in the protocol implementations as they are. OTOH, protocol implementations that are not interested in using 103 can set `send_early_hints` to NULL so that the `h2o_send_early_hints` function can bail out earlier.

> Do you mean we should have a configuration knob which allows the users to turn off 103 forwarding only when HTTP/1 is used?

I had thought that we had agreed to a knob that turns of sending 103 in any case for H1 (or both H1 and H2). Was that a discussion of a different issue?